subprojects/android-test/ui-testing-core/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt

```diff
@@ -0,0 +1,74 @@
+package com.avito.android.rule
+
+import android.app.Activity
+import android.app.Instrumentation
+import android.content.Intent
+import androidx.annotation.CallSuper
+import androidx.test.core.app.ActivityScenario
+import androidx.test.platform.app.InstrumentationRegistry
+import org.junit.rules.ExternalResource
+import java.lang.ref.WeakReference
+
+/**
+ * This rule provides functional testing of a single [Activity] similar to deprecated
+ * [androidx.test.rule.ActivityTestRule].
+ * @param activityClass The Activity class under test
+ * @param launchActivity true if the Activity should be launched before each test
+ */
+open class ActivityScenarioRule<A : Activity>(
+    private val activityClass: Class<A>,
+    private val launchActivity: Boolean
+) : ExternalResource() {
+
+    val activity: A
+        get() = activityRef?.get() ?: throwUninitializedException()
+    val activityResult: Instrumentation.ActivityResult
+        get() = scenario?.result ?: throwUninitializedException()
+
+    private var scenario: ActivityScenario<A>? = null
+    private var activityRef: WeakReference<A>? = null
+
+    override fun before() {
+        super.before()
+        if (launchActivity) {
+            scenario = ActivityScenario.launch(activityClass)
+            afterActivityLaunched()
+        }
+    }
+
+    override fun after() {
+        super.after()
+        scenario?.run {
+            close()
+            afterActivityFinished()
+        }
+        activityRef = null
+    }
+
+    fun launchActivity(intent: Intent?): A {
+        scenario = when (intent) {
+            null -> ActivityScenario.launch(activityClass)
+            else -> ActivityScenario.launch(
+                intent.setClass(InstrumentationRegistry.getInstrumentation().targetContext, activityClass)
+            )
+        }
+        afterActivityLaunched()
+        return activityRef?.get() ?: throwUninitializedException()
+    }
+
+    fun runOnUiThread(runnable: Runnable) {
+        scenario?.onActivity {
+            runnable.run()
+        } ?: error("Activity is not launched")
+    }
+
+    private fun throwUninitializedException(): Nothing = error("Activity is not initialized")
```

Comment:
```suggestion
    private inline fun throwUninitializedException()= error("Activity is not initialized")
```

Reply:
Let's also add `activityClass` to error message

Reply:
`Nothing` return type must be always specified explicitly. Also I'm not sure if inlining is necessary here.

Reply:
Added `activityClass` to error message